runtime/protocol/http-client-engines/http-client-engine-crt/common/src/aws/smithy/kotlin/runtime/http/engine/crt/CrtHttpEngine.kt

```diff
@@ -168,3 +144,32 @@ public class CrtHttpEngine(public val config: CrtHttpEngineConfig) : HttpClientE
         }
     }
 }
+
+internal suspend fun HttpStream.sendChunkedBody(body: HttpBody) {
+    when (body) {
+        is HttpBody.SourceContent -> {
+            val source = body.readFrom()
+            val bufferedSource = source.buffer()
+
+            while (!bufferedSource.exhausted()) {
+                bufferedSource.request(CHUNK_BUFFER_SIZE)
+                this.writeChunk(bufferedSource.buffer.readByteArray(), isFinalChunk = bufferedSource.exhausted())
+            }
+        }
+        is HttpBody.ChannelContent -> {
+            val chan = body.readFrom()
+            var buffer = SdkBuffer()
+            val nextBuffer = SdkBuffer()
+
+            while (!chan.isClosedForRead) {
+                chan.read(buffer, CHUNK_BUFFER_SIZE)
+
+                val isFinalChunk = chan.read(nextBuffer, CHUNK_BUFFER_SIZE) == -1L
+
+                this.writeChunk(buffer.readToByteArray(), isFinalChunk)
+                if (isFinalChunk) break else buffer = nextBuffer
+            }
+        }
+        else -> {}
```

Comment:
**Nit**: `else -> error("sendChunkedBody shouldn't be called for non-chunked body types")` or something like that.